File: assets/old/PPTXMerger.scala
package inosion.pptx


import org.apache.poi.xslf.usermodel._ // XMLSlideShow
import org.apache.poi.sl.usermodel._
import java.io.File
import java.io._

import scala.collection.JavaConverters._
import com.fasterxml.jackson.databind.JsonNode


import org.openxmlformats.schemas.drawingml.x2006.main.CTTableRow

import io.gatling.jsonpath.JsonPath

import scala.util._
import scala.collection.mutable
import org.apache.poi.sl.usermodel.PictureData.PictureType
import java.awt.geom.Rectangle2D
import com.typesafe.scalalogging._
//import org.openxmlformats.schemas.presentationml.x2006.main.


object ShapeGroup extends Enumeration  {
  type ShapeGroup = Value
  val Image = Value
}


// Value holder for data context
case class DataContext(rootJsonNode: JsonNode, contextJsonNode: Option[JsonNode] = None)


object PPTXMerger extends StrictLogging {

    type TextHolder = TextShape[XSLFShape, XSLFTextParagraph]

    type Foo = XSLFSheet

    import scala.reflect.ClassTag

    def f[T](v: T)(implicit ev: ClassTag[T]) = ev.toString

    val JsonPathReg                   = raw"[\$$@]|[\$$@]\.[0-9A-Za-z_:\-\?\.\[\]\*]+"
    val matchRegexpTemplate           = raw".*(\{\{\s*(" + JsonPathReg + raw")\s*\}\}).*"
    val matchContextControl           = (raw"(\{\s*context\s*=\s*(" + JsonPathReg + raw")\s*\})").r
    val matchGroupShapeControl        = (raw"(\{\s*context\s*=\s*(" + JsonPathReg + raw")\s*dir\s*=\s*(\d+)\s+gap=(\d+)\s*\})").r

    // new methods for iteration here

    /**
     * Iterates over the source template slide.
     */
    def render2(fc: FilesContext): Either[Error, Unit] = {

        // we need to track which SlideLayouts we ported across
        val visitedLayouts: mutable.Seq[XSLFSlideLayout] = mutable.Seq()
        logger.debug(s"! Regexp for templates = ${matchRegexpTemplate}")
        logger.debug(s"! Regexp for controls = ${matchContextControl}")
        logger.debug(s"! Regexp for group shape controls = ${matchGroupShapeControl}")

        val pptTemplate: XMLSlideShow = new XMLSlideShow(new FileInputStream(fc.src.getAbsolutePath()))
        val pptNew: XMLSlideShow      = new XMLSlideShow(new FileInputStream(fc.src.getAbsolutePath()))

        // To retain the Master Slides, fonts etc... we start with the original Slide and
        // to retain the master slide layouts.
        for (idx <- 0 until pptNew.getSlides.size()) {
            logger.debug(s"√ Removing slide ${prettyPrintSlide(pptNew.getSlides().get(0))}")
            pptNew.removeSlide(0)
        }

        // get the data
        val jsonData = JsonYamlTools.readFileToJson(fc.data)

        // iterate over the old slide
        for (srcSlide <- pptTemplate.getSlides().asScala) {
            val sc = SlidesContext(srcSlide = srcSlide, destPptx = pptNew, destSlides = List())
            processSlide(fc, sc, dc = DataContext(rootJsonNode = jsonData), visitedLayouts)
        }

        // save out the resulting new pptx
        try {
            pptNew.write(new FileOutputStream(fc.dst))
            Right(())
        } catch {
            case e: Exception => Left(new Error(s"failed writing the new pptx [${fc.dstAbsPath}]", e))
        }

    }

    def processSlide(fc: FilesContext, sc: SlidesContext, dc: DataContext, visitedLayouts: mutable.Seq[XSLFSlideLayout]) {

        // We will copy this slide from source to new ppt, then fill it's data
        val newSlide = PPTXTools.createSlide(sc.destPptx, sc.srcSlide, visitedLayouts)
        val newSlidesContext = sc.copy(destSlides = sc.destSlides :+ newSlide)

        System.err.println(s":: New slide from [${fc.srcRelPath}(${prettyPrintSlide(sc.srcSlide)})] to [${fc.dstRelPath}(${prettyPrintSlide(sc.srcSlide)})] ")

        findControlJsonPath(sc.srcSlide.getShapes().asScala) match {
            case Some(PageControlData(shape, contextMatch, jsonPath)) => {
                logger.debug(s"√ Slide [${prettyPrintSlide(sc.srcSlide)}] has a jsonPath context -> ${jsonPath}")
                JsonPath.query(jsonPath, dc.rootJsonNode).map{ i =>
                    for (jsonNode <- i) {
                        logger.debug(s"√ node = ${jsonNode.toString()}")
                        processSlide(fc, newSlidesContext, dc.copy(contextJsonNode = Some(jsonNode)), visitedLayouts)
                    }
                }
                // now we have templated it out, let's remove it
                sc.destPptx.removeSlide(newSlide.getSlideNumber() - 1)
                newSlide.removeShape(shape)
            }
            case None => processAllShapes(newSlide, dc)
        }

    }

    def processAllShapes(slide: XSLFSlide,  dc: DataContext) {

        for (shape <- slide.getShapes().asScala) {

            shape match {
                case textShape : TextHolder =>
                    if (hasTemplate(textShape)) {
                      logger.debug(s"√ ${f(shape)} -> ${textShape.getText()}")
                      changeText(textShape, dc)
                      logger.debug(s"√ now == ${f(shape)} -> ${textShape.getText()}")

                    } else {
                      logger.debug(s"✖ '${textShape.getText()}' did not match")
                    }

                case group : XSLFGroupShape => processGroupShape(group)
                case table : XSLFTable  if (hasControl(table.getRows().get(0).getCells().get(0)))    => {
                    logger.debug(s"√ we have a table - ${table.getRows().get(0).getCells().get(0).getText()}")
                    iterateTable(table, dc)
                }
                case _ => logger.debug(s"✖ ${f(shape)} is not TextHolder")
            }

        }
    }

    def processGroupShape(groupShape: XSLFGroupShape) {
        logger.debug(s"⸮ Inspecting XSLFGroupShape[${groupShape.getShapeName()}]...")
        findControlJsonPath(groupShape.getShapes().asScala) match {
            case None => logger.debug(s"✖ XSLFGroupShape[${groupShape.getShapeName()}] - no control, ignoring")
            case Some(GroupShapeControlData(shape, controlMatch, jsonPath, direction, gap)) => {
                logger.debug(s"√ Found the XSLFGroupShape[${groupShape.getShapeName()}] with control fields")
                // from here we get a Exception in thread "main" java.util.ConcurrentModificationException
                // val newGroupShape: XSLFGroupShape = groupShape.getSheet().createGroup()
                // newGroupShape.setAnchor(groupShape.getAnchor())

                for (shape <- groupShape.getShapes().asScala) {
                    shape match {
                        case s: XSLFAutoShape => logger.debug("not impl yet") // newGroupShape.createAutoShape()
                        case t: XSLFTextBox   => logger.debug("groupshape - not yet") // newGroupShape.createTextBox().setText(t.getText())
                        case _ => logger.error(s"The group ${prettyPrintShape(groupShape)} has a shape ${prettyPrintShape(shape)} that did not match")
                    }

                }
                // nned to remove it on the outer /// groupShape.removeShape(shape)
            }
        }
    }

    def iterateTable(table: XSLFTable, dc: DataContext) {
        val firstCell = table.getRows().get(0).getCells().get(0)
        val firstCellText = firstCell.getText()
        // val (a, b, tableContextJsonPath) = for (m <- matchContextControl.findFirstMatchIn(firstCellText)) yield m.group
        val mm = for (m <- matchContextControl.findFirstMatchIn(firstCellText)) yield m
        val tableContextJsonPath = mm.get.group(2)
        val controlString = mm.get.group(0)

        val (jsonNode, _tableContextJsonPath) = nodeAndQuery(tableContextJsonPath, dc)

        JsonPath.query(_tableContextJsonPath, jsonNode).map{ iter =>
            for ((jsonNode, i) <- iter.zipWithIndex) {
                logger.debug(s"√ Table node = ${jsonNode.toString()}")
                RowCloner.cloneRow(table, 1) // including cells
                for ((cell, ci) <- table.getRows().get(table.getRows().size() - 1).getCells().asScala.zipWithIndex) {
                    // cell.setText(r)
                    cell.setStrokeStyle(cell.getStrokeStyle())
                       // get the current style and size etc
                       // src cell
                    val srcCell = table.getRows().get(0).getCells().get(ci)
                    val newText = replaceText(dc.copy(contextJsonNode = Some(jsonNode)), cell.getText())
                    applyNewText(newText, cell, srcCell)
                }
            }
        }

        // now remove the template row
        table.removeRow(1)
        // remove the context string
        //table.getRows().get(0).getCells().get(0).setText(firstCellText.replace(controlString, ""))


        applyNewText(firstCellText.replace(controlString, ""), firstCell, firstCell)



    }

    /**
     * Replaces the text in the "textShape", cloning the style out of the "first" paragraph textrun
     */
    def applyNewText(newText: String, textHolder: XSLFTextShape, original: XSLFTextShape) {
        logger.debug(s"√ applyNewText :: [${textHolder.getText()}] newText = [${newText}]")
        val color = original.getTextParagraphs().get(0).getTextRuns().get(0).getFontColor()
        val font = original.getTextParagraphs().get(0).getTextRuns().get(0).getFontFamily()
        val size = original.getTextParagraphs().get(0).getTextRuns().get(0).getFontSize()

        textHolder.setText("") // effectively removes all paragraphs and textruns

        val tr:XSLFTextRun = textHolder.appendText(newText,true)
        tr.setFontColor(color)
        tr.setFontFamily(font)
        tr.setFontSize(size)


    }

    def changeText(textShape: TextHolder, dc: DataContext) {
        val text = textShape.getText()
        val matchRegexpTemplate.r(replacingText, jsonQuery) = text

        logger.debug(s"√ found = [${replacingText}] jsonpath = [${jsonQuery}]")

        val (theJsonNode, theJsonPath) = nodeAndQuery(jsonQuery, dc)

        val dataText = JsonPath.query(theJsonPath, theJsonNode) match {
            case Left(error) => throw new java.lang.Error(error.reason)
            case Right(i)    => try {
                Some(i.next().asText())
            } catch {
                case e: java.util.NoSuchElementException => logger.error(s"The JSONPath expression ==> ${theJsonPath} <== did not resolve to any data. Ignoring"); None
                case e: Exception => throw e
            }
        }

        dataText.map { txt =>
            val newText = text.replace(replacingText, txt)

            applyNewText(newText, textShape.asInstanceOf[XSLFTextShape], textShape.asInstanceOf[XSLFTextShape])

            /*
            // get the current style and size etc
            val color = textShape.getTextParagraphs().get(0).getTextRuns().get(0).getFontColor()
            val font = textShape.getTextParagraphs().get(0).getTextRuns().get(0).getFontFamily()
            val size = textShape.getTextParagraphs().get(0).getTextRuns().get(0).getFontSize()
            logger.debug(s"√ dataText = [${txt}] newText = [${newText}]")

            textShape.setText("")
            val tr:XSLFTextRun = textShape.appendText(newText,true).asInstanceOf[XSLFTextRun]
            tr.setFontColor(color)
            tr.setFontFamily(font)
            tr.setFontSize(size)
            */
        }
    }

    def replaceText(dc: DataContext, text: String) :String = {
        // fullTemplateText is {{someJsonPath}}
        // jsonQuery is what was found
        val matchRegexpTemplate.r(fullTemplateText, jsonQuery) = text

        val (jsonNode, jsonPath) = nodeAndQuery(jsonQuery, dc)

        val dataText = JsonPath.query(jsonPath, jsonNode) match {
            case Left(error) => throw new java.lang.Error(error.reason)
            case Right(i)    => try {
                Some(i.next().asText())
            } catch {
                case e: java.util.NoSuchElementException => System.err.println(s"The JSONPath expression ==> ${jsonQuery} <== did not resolve to any data. Ignoring"); None
                case e: Exception => throw e
            }
        }

        dataText.map( text.replace(fullTemplateText, _) ).getOrElse("")

    }

    /**
     * Given a JSONPath query ($. or @.) we will determine
     * if it is to use the context node [@.], or the rootnode [$.]
     * for the lookup
     *
     * ! This is a hack because the io.gatling.JsonPath does not support a context object.
     */
    def nodeAndQuery(jsonQuery: String, dc: DataContext): (JsonNode, String) = {
        jsonQuery(0) match {
            case '$' => (dc.rootJsonNode,        jsonQuery)
            case '@' => dc.contextJsonNode match {
                            case Some(jn) => (jn, "$" + jsonQuery.stripPrefix("@"))
                            case None     => {
                                val newJsonPath = "$" + jsonQuery.stripPrefix("@")
                                logger.warn(s"! jsonPath starts is ${jsonQuery} but the context object is empty. Using root object instead (eg: ${newJsonPath})")
                                (dc.rootJsonNode, "$" + jsonQuery.stripPrefix("@"))
                            }
                        }
        }
    }

    def addImage(ppt: XMLSlideShow, slide: XSLFSlide, imagePath: String, imageShapeName: String, shape: XSLFShape, pictureType: PictureType) {

        val picIS: FileInputStream = new FileInputStream(new File(imagePath))
        // https://stackoverflow.com/questions/4905393/scala-inputstream-to-arraybyte commons-io still the best
        val picture: Array[Byte]       = org.apache.commons.io.IOUtils.toByteArray(picIS)

        val anchor: Rectangle2D = shape.getAnchor()
        slide.removeShape(shape)

        val pd: XSLFPictureData    = ppt.addPicture(picture, pictureType)
        val pics: XSLFPictureShape = slide.createPicture(pd)
        pics.setAnchor(anchor)

    }

    trait ControlData {
        // the shape that holds the matching control string (for removal by caller)
        def shape: XSLFShape
        // the string that matched
        def controlText: String
        // the jsonPath
        def jsonPath: String
    }
    case class PageControlData(shape: XSLFShape, controlText: String, jsonPath: String) extends ControlData
    case class GroupShapeControlData(shape: XSLFShape, controlText: String, jsonPath: String, direction: Int, gap: Int) extends ControlData
    case class ImageControlData(shape: XSLFShape, controlText: String, jsonPath: String) extends ControlData


    def findControlJsonPath(shapes: scala.collection.mutable.Seq[XSLFShape]): Option[ControlData] = {

        for (shape <- shapes) {
            if (shape.isInstanceOf[TextHolder]) {
                val textShape = shape.asInstanceOf[TextHolder]
                logger.debug(s"⸮ inspecting - Shape[${shape.getShapeName()}] `${textShape.getText()}`")
                textShape.getText() match {
                    case matchContextControl(controlText, jsonPath) => { // page control
                        logger.debug(s"√ Match - Shape[${shape.getShapeName()}] control=`${controlText}` jp=`${jsonPath}`")
                        return Some(PageControlData(shape, controlText, jsonPath))
                    }
                    case matchGroupShapeControl(controlText, jsonPath, direction, gap) => {
                        logger.debug(s"√ Match - Shape[${shape.getShapeName()}] control=`${controlText}` jp=`${jsonPath}` dir=${direction} gap=${gap}")
                        return Some(GroupShapeControlData(shape, controlText, jsonPath, direction.toInt, gap.toInt))
                    }
                    case _ => {
                        logger.debug(s"✖ shape:${shape.getShapeName()} did not have a controlData")

                    }
                }
            }
        }
        None
    }

    def getShape(slide: XSLFSlide, shapeName: String): Option[XSLFShape] = {
        for (shape <- slide.getSlideLayout().getShapes().asScala) {
            shape.getShapeName().toLowerCase() match {
              case shapeName => return Some(shape)
              case _ =>
            }
        }
        return None;
    }

    def hasTemplate(shape: TextHolder): Boolean =
      shape.getText().matches(matchRegexpTemplate)

    def hasControl(shape: TextHolder): Boolean = matchContextControl.findFirstIn(shape.getText()).isDefined


    def prettyPrintSlide(slide: XSLFSlide): String = {
        List(Option(slide.getSlideName()), Some(s"#${slide.getSlideNumber()}")).flatten.mkString("/")
    }

    def prettyPrintShape(shape: XSLFShape): String = shape.getShapeName()


}

File: src/pptx_merger.py
import re
import json
import logging
from pptx import Presentation
from pptx.util import Inches
from pptx.enum.shapes import MSO_SHAPE_TYPE
from pptx.enum.shapes import PP_PLACEHOLDER
from pptx.dml.color import RGBColor

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)

class PPTXMerger:
    JsonPathReg = r"[\$@]|[\$@]\.[0-9A-Za-z_:\-\?\.\[\]\*]+"
    matchRegexpTemplate = re.compile(r".*(\{\{\s*(?P<jsonpath>" + JsonPathReg + r")\s*\}\}).*")
    matchContextControl = re.compile(r"(\{\s*context\s*=\s*(?P<jsonpath>" + JsonPathReg + r")\s*\})")
    _grpShape = re.compile(r"(?s)(\{\s*context\s*=\s*(?P<jsonpath>" + JsonPathReg + r")\s*dir\s*=\s*(?P<dir>\d+)(\s+gap=(?P<gap>\d+))?\s*\})")

    @staticmethod
    def render(data_file, template_file, out_file):
        visited_layouts = []
        logger.debug(f"! Regexp for templates = {PPTXMerger.matchRegexpTemplate}")
        logger.debug(f"! Regexp for controls = {PPTXMerger.matchContextControl}")
        logger.debug(f"! Regexp for group shape controls = {PPTXMerger._grpShape}")

        ppt_template = Presentation(template_file)
        ppt_new = Presentation(template_file)

        # Remove all slides to retain the master slide layouts
        while ppt_new.slides:
            slide = ppt_new.slides[0]
            logger.debug(f"√ Removing slide {PPTXMerger.pretty_print_slide(slide)}")
            ppt_new.slides.remove(slide)

        with open(data_file, 'r') as f:
            json_data = json.load(f)

        for src_slide in ppt_template.slides:
            PPTXMerger.process_slide(template_file, out_file, src_slide, ppt_new, json_data, None, visited_layouts)

        ppt_new.save(out_file)
        return

    @staticmethod
    def pretty_print_slide(slide):
        return f"{slide.name} #{slide.slide_id}"

    @staticmethod
    def pretty_print_shape(shape):
        return shape.name

    @staticmethod
    def process_slide(src_path, dest_path, src_slide, ppt_new, root_json, context_json, visited_layouts):
        new_slide = PPTXTools.create_slide(ppt_new, src_slide, visited_layouts)
        logger.debug(f":: ------ New Slide from src [{src_path}({src_slide.slide_id})] to [{dest_path}({new_slide.slide_id})] ")

        control_data = PPTXMerger.find_control_json_path(new_slide.shapes)
        if control_data:
            if isinstance(control_data, PageControlData):
                logger.debug(f"√ Slide [{PPTXMerger.pretty_print_slide(new_slide)}] has a PageControl - jsonPath context -> {control_data.json_path}")
                new_slide.shapes._spTree.remove(control_data.shape._element)
                result = jsonpath_rw.parse(control_data.json_path).find(root_json)
                for match in result:
                    PPTXMerger.process_slide(src_path, dest_path, new_slide, ppt_new, root_json, match.value, visited_layouts)
                ppt_new.slides.remove(new_slide)
            else:
                logger.error(f"! Slide [{PPTXMerger.pretty_print_slide(new_slide)}] has a control but it is not a PageControlData")
        else:
            PPTXMerger.process_all_shapes(new_slide, root_json, context_json)

    @staticmethod
    def process_all_shapes(slide, root_json, context_json):
        for shape in slide.shapes:
            if shape.shape_type == MSO_SHAPE_TYPE.GROUP:
                PPTXMerger.process_group_shape(shape)
            elif shape.has_text_frame:
                if PPTXMerger.has_template(shape):
                    logger.debug(f"√ {shape} is a templated shape")
                    PPTXMerger.change_text(shape, root_json, context_json)
                else:
                    logger.debug(f"✖ '{shape.text}' did not match")
            elif shape.shape_type == MSO_SHAPE_TYPE.TABLE:
                if PPTXMerger.has_control(shape.table.cell(0, 0)):
                    logger.debug(f"√ we have a table - {shape.table.cell(0, 0).text}")
                    PPTXMerger.iterate_table(shape.table, root_json, context_json)
            else:
                logger.debug(f"✖ {shape} is not TextHolder")

    @staticmethod
    def process_group_shape(group_shape):
        logger.debug(f"⸮ Inspecting GroupShape[{group_shape.name}]...")
        control_data = PPTXMerger.find_control_json_path(group_shape.shapes)
        if control_data:
            if isinstance(control_data, GroupShapeControlData):
                logger.debug(f"√ Found the GroupShape[{group_shape.name}] with control fields")
                new_group_shape = group_shape.parent.shapes.add_group_shape()
                new_group_shape.left = group_shape.left
                new_group_shape.top = group_shape.top
                new_group_shape.width = group_shape.width
                new_group_shape.height = group_shape.height
                # Process shapes within the group
                for shape in group_shape.shapes:
                    if shape.shape_type == MSO_SHAPE_TYPE.AUTO_SHAPE:
                        new_shape = new_group_shape.shapes.add_shape(shape.auto_shape_type, shape.left, shape.top, shape.width, shape.height)
                        new_shape.text = shape.text
                    elif shape.shape_type == MSO_SHAPE_TYPE.GROUP:
                        PPTXMerger.process_group_shape(shape)
                    elif shape.shape_type == MSO_SHAPE_TYPE.TEXT_BOX:
                        new_shape = new_group_shape.shapes.add_textbox(shape.left, shape.top, shape.width, shape.height)
                        new_shape.text = shape.text
                    else:
                        logger.error(f"The group {PPTXMerger.pretty_print_shape(group_shape)} has a shape {PPTXMerger.pretty_print_shape(shape)} that did not match")
            group_shape._spTree.remove(control_data.shape._element)
        else:
            logger.debug(f"✖ GroupShape[{group_shape.name}] - no control, ignoring")

    @staticmethod
    def iterate_table(table, root_json, context_json):
        first_cell_text = table.cell(0, 0).text
        match = PPTXMerger.matchContextControl.search(first_cell_text)
        if match:
            table_context_json_path = match.group("jsonpath")
            control_string = match.group(0)
            json_node, table_context_json_path = PPTXMerger.node_and_query(table_context_json_path, root_json, context_json)
            result = jsonpath_rw.parse(table_context_json_path).find(json_node)
            for match in result:
                new_row = table.add_row()
                for cell, json_node in zip(new_row.cells, match.value):
                    PPTXMerger.change_text(cell, root_json, json_node)
                    cell.fill.solid()
                    cell.fill.fore_color.rgb = RGBColor(255, 255, 255)
            table.rows[1]._element.getparent().remove(table.rows[1]._element)
            table.cell(0, 0).text = first_cell_text.replace(control_string, "")

    @staticmethod
    def change_text(shape, root_json, context_json):
        text = shape.text
        match = PPTXMerger.matchRegexpTemplate.search(text)
        if match:
            replacing_text = match.group(1)
            json_query = match.group("jsonpath")
            logger.debug(f"√ found = [{replacing_text}] jsonpath = [{json_query}]")
            json_node, json_path = PPTXMerger.node_and_query(json_query, root_json, context_json)
            result = jsonpath_rw.parse(json_path).find(json_node)
            for match in result:
                new_text = text.replace(replacing_text, match.value)
                logger.debug(f"√ dataText = [{match.value}] newText = [{new_text}]")
                shape.text = new_text

    @staticmethod
    def node_and_query(json_query, root_json, context_json):
        if json_query.startswith('$'):
            return root_json, json_query
        elif json_query.startswith('@'):
            if context_json:
                return context_json, '$' + json_query[1:]
            else:
                new_json_path = '$' + json_query[1:]
                logger.warn(f"! jsonPath starts is {json_query} but the context object is empty. Using root object instead (eg: {new_json_path})")
                return root_json, new_json_path

    @staticmethod
    def find_control_json_path(shapes):
        for shape in shapes:
            if shape.has_text_frame:
                text = shape.text
                logger.debug(f"⸮ inspecting - Shape[{shape.name}] `{text}`")
                match = PPTXMerger._grpShape.search(text)
                if match:
                    json_path = match.group("jsonpath")
                    direction = match.group("dir")
                    gap = int(match.group("gap")) if match.group("gap") else 0
                    control_text = match.group(0)
                    logger.debug(f"√ Match - Shape[{shape.name}] control=`{control_text}` jp=`{json_path}` dir={direction} gap={gap}")
                    return GroupShapeControlData(shape, control_text, json_path, int(direction), gap)
                match = PPTXMerger.matchContextControl.search(text)
                if match:
                    json_path = match.group("jsonpath")
                    control_text = match.group(0)
                    logger.debug(f"√ Match - Shape[{shape.name}] control=`{control_text}` jp=`{json_path}`")
                    return PageControlData(shape, control_text, json_path)
                logger.debug(f"✖ shape:{shape.name} {text} did not have a controlData")
        return None

    @staticmethod
    def has_template(shape):
        return PPTXMerger.matchRegexpTemplate.search(shape.text) is not None

    @staticmethod
    def has_control(shape):
        return PPTXMerger.matchContextControl.search(shape.text) is not None


class ControlData:
    def __init__(self, shape, control_text, json_path):
        self.shape = shape
        self.control_text = control_text
        self.json_path = json_path


class PageControlData(ControlData):
    pass


class GroupShapeControlData(ControlData):
    def __init__(self, shape, control_text, json_path, direction, gap):
        super().__init__(shape, control_text, json_path)
        self.direction = direction
        self.gap = gap


class ImageControlData(ControlData):
    pass


class PPTXTools:
    @staticmethod
    def create_slide(prs, src_slide, visited_layouts):
        slide_layout = src_slide.slide_layout
        if slide_layout not in visited_layouts:
            visited_layouts.append(slide_layout)
        slide = prs.slides.add_slide(slide_layout)
        PPTXMerger.copy_slide_content(src_slide, slide)
        return slide

    @staticmethod
    def copy_slide_content(src_slide, dest_slide):
        for shape in src_slide.shapes:
            if shape.has_text_frame:
                new_shape = dest_slide.shapes.add_textbox(shape.left, shape.top, shape.width, shape.height)
                new_shape.text = shape.text
            elif shape.shape_type == MSO_SHAPE_TYPE.PICTURE:
                image_stream = shape.image.blob
                dest_slide.shapes.add_picture(image_stream, shape.left, shape.top, shape.width, shape.height)
            # Add more shape types as needed